README.md

```diff
@@ -92,14 +92,219 @@ class Genome(ABC):
 
-You are free to implement the genome classes however you want, and using whateer auxilary data structures you desire, as long as one uses a Python list with an element for each nucleotide and the other a linked list with a link for each nucleotide. If you want to implement a third (or fourth or fifth...) version, you are very welcome to do so as well.
+You are free to implement the genome classes however you want, and use whatever auxiliary data structures you desire, as long as one uses a Python list with an element for each nucleotide and the other a linked list with a link for each nucleotide. If you want to implement a third (or fourth or fifth...) version, you are very welcome to do so as well.
 
 ## Complexity
 
 When you have implemented the two (or more) classes, describe the complexity of each operation as a function of the genome size (at the time of the operation), and the size of the TE involved (and when copying, the offset you are copying). Put the description here:
 
-**FIXME: OPERATION COMPLEXITY**
+## List implementation
 
-In `src/simulate.py` you will find a program that can run simulations and tell you actual time it takes to simulate with different implementations. You can use it to test your analysis. You can modify the parameters to the simulator if you want to explore how they affect the running time.
+During this exercise, we will use $n$ as the length of the genome, $m$ as the length of the transposable element, and $k$ as the number of "blocks". 
+
+First, let's start with the List implementation. 
+
+### Init instance
+
+```python
+def __init__(self, n: int):
+    self.genome = n*['-'] # O(n)
+    self.active_identifiers = {} # O(1)
+    self.te_counter = 0# O(1)
+```
+
+Since appending to a list has a complexity of $O(1)$, creating the genome costs $O(n)$, whereas $n$ is the length of the initial genome. 
+
+### Insert a transposable element to the list
+
+```python
+def insert_te(self, pos: int, length: int) -> int:
+    pos = pos % len(self) # O(1)
+    active_identifiers = list(self.active_identifiers.items()) # O(k)
+    for identifier,active_te  in active_identifiers: # O(k)
+        is_in_range(pos, active_te): # O(1)
+            start, end = active_te # O(1)
+            self.genome[start:end] = (end - start)*["x"] # O(n + m)
+            self.active_identifiers.pop(identifier) # O(1)
+        if active_te.start > pos: O(1)
+            self.active_identifiers[identifier] = Interval(
+                active_te.start + length, active_te.end + length
+            ) # O(1)
+    self.genome[pos:pos] = length*['A'] # O(n + m)
+    self.te_counter += 1 # O(1)
+    self.active_identifiers[self.te_counter] = Interval(pos, pos + length) # O(1)
+    return self.te_counter
+```
+
+As you can see in the previous block, the final complexity is $O(k(n+m))$. However, in the worst case, all transposable elements are of length one and $k = n$, then we have $O(n^2\cdot m)$
```

Comment:
Your results are correct but overly pessimistic. It is only $O(k + n + m)$.

You make a common mistake when reasoning about nested loops: you multiply the complexity of the inner loop with how many times you run the outer loop. That will always give you a correct upper bound, but it will be overly pessimistic if you don't actually _run_ the inner loop for every iteration of the outer. When you insert a new TE, exactly one (active) TE can sit at the position you are inserting it on. (If you have inactive elements in `active_identifiers` then you get what you deserve, and I can't help you with that kind of thinking, but if you are sensible, as I believe you are, then only one TE can be hit). That means that the inner loop (the slice update to the genome) will only happen once.

So if we separate the complexity of moving positions for all active TEs, something you do in $O(k)$, from the single time you update a hit TE in time $O(n+m)$, we get $O(k+n+m)$.

If you stored the TE identifies in `self.genome` instead of strings, you could avoid the search through the active TEs as well. You don't need to have strings in `self.genome` (you just need to generate them when you call `__str__`), so they can contain IDs instead. Then you are down to $O(n+m)$. That is as good as it gets with an array implementation.